keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+* `kube_pod_resource_requests_lifecycle` contains series for the consumed requested resources for a pod at a given time when the pod is in the provided lifecycle phase. All pods are in one of the lifecycle phases, but series will only be returned for the current phase.
+  * The value is the quantity of the limit in `spec.containers.resources.requests`
+  * It has the following labels:
+    * `lifecycle` - this is set to the current lifecycle phase as described in the resource model - `Pending`, `Initializing`, `Running`, `Completed`
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.requests`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.requests`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+* `kube_pod_resource_limits_lifecycle` contains series for the consumed limit resources for a pod at a given time when the pod is in the provided lifecycle phase.
+All pods are in one of the lifecycle phases, but series will only be returned for the current phase.
+  * The value is the quantity of the limit in `spec.containers.resources.limits`
+  * It has the following labels:
+    * `lifecycle` - this is set to the current lifecycle phase as described in the resource model - `Pending`, `Initializing`, `Running`, `Completed`
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.limits`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+
+This will add `O(pods * resource types)` metrics series, and in general most clusters have 4-8 resource types. The series per pod is `resource_count * 2 * 2 (requests and limits for normal and _lifecycle)`.  This explicitly does not expose containers as a dimension, as the resource model applies rules to containers that cannot be trivially expressed via a formula and does not help answer the core questions.
+
+These metrics should roughly correspond to:
+
+* The resources the scheduler will consider the pod to be consuming at the current time
+* The resources the kubelet will use to decide whether to admit the pod (nodes will reject requested resources that exceed their actual allocatable capacity)
+* The effective reserved capacity as resource quota would observe for the namespace
+* The sum of all resources consumed on the cluster
+* The size of cgroup constraints applied to pods on a node
+
+The lifecycle metrics allow key insights that are impossible today:
```

Comment:
nit: I'm not sure "Impossible" is accurate.  I'd propose "would require a join" instead.

Reply:
Given how even internal kube contributors get this wrong, I'm actually going to say that anything we don't expose as first class is wrong in multiple places and confusing end users :)  Requiring a join here is hard because you don't even know to ask about the join.  The number of incorrect dashboards providing business critical insight based on flawed or incorrect assumptions about the resource model is fairly wide, and even 5 years in I see it consistently regressed.

I'd like to tighten the bounds on the system so that a user seeing this model is consistently exposed to the concept, not just the mechanical fields.

Reply:
I'm not disagreeing, just making sure other readers can understand and evaluate the tradeoffs.

Reply:
Absolutely - I meant "I am going to change language, but do you agree that understanding the model is worth exposing on its own?" and also "what are some other concrete things we could do to improve this"

Reply:
Got it.  I'm undecided on this particular point, and am working towards agreement in other threads.  If I don't think something is workable, i'll try and understand your goals so I can propose an alternative.

Reply:
Suggestion: s/impossible/difficult/

Reply:
+1 on writing.